core/src/main/java/org/wso2/carbon/kernel/internal/startupresolver/beans/StartupComponent.java

```diff
@@ -118,24 +123,62 @@ public boolean isServiceRequired(String service) {
     /**
      * Register {@code Capability} instances with this startup listener component.
      * <p>
-     * This {@code Capability} may be in the AVAILABLE or EXPECTED state.
-     * <p>
-     * 1) If the {@code Capability} is in the EXPECTED
-     * state and if there is matching {@code Capability} already registered with this startup listener component, then
-     * it is removed from the list of {@code Capability} of this component.
-     * <p>
-     * 2) If the {@code Capability} is in the AVAILABLE
-     * state and if there is matching {@code Capability} already registered with this startup listener component, then
-     * it is removed from the list of {@code Capability} of this component.
+     * Adds the given {@code Capability} to the {@code expectedCapabilityList}
      *
      * @param capability {@code Capability} object to be registered with this startup listener component.
      */
-    public void addExpectedOrAvailableCapability(Capability capability) {
-        synchronized (pendingCapabilityList) {
-            if (pendingCapabilityList.contains(capability)) {
-                pendingCapabilityList.remove(capability);
+    public void addExpectedCapability(Capability capability) {
+        synchronized (expectedCapabilityList) {
+            Capability expectedCapability = expectedCapabilityList.stream()
+                    .filter(cap -> cap.getName().equals(capability.getName()))
+                    .filter(cap -> !cap.isSecondCheck()
+                            && cap.getState() == Capability.CapabilityState.AVAILABLE)
+                    .findFirst().orElse(null);
+
+            if (expectedCapability != null) {
+                expectedCapability.setSecondCheck(true);
+                expectedCapability.setDirectDependency(capability.isDirectDependency());
             } else {
-                pendingCapabilityList.add(capability);
+                expectedCapabilityList.add(capability);
+            }
+        }
+    }
+
+    /**
+     * This method updates the capability in the expected capability list.
+     *
+     * If a corresponding capability is found in the expectedCapabilityList, then the existing capability is updated,
+     * or the new capability is added to the expectedCapabilityList otherwise.
+     *
+     * @param capability the capability to be updated
+     */
+    public void updateCapability(Capability capability) {
+        synchronized (expectedCapabilityList) {
+            if (capability.getState() == Capability.CapabilityState.EXPECTED) {
+                Optional<Capability> optCapability = expectedCapabilityList.stream()
+                        .filter(cap -> cap.getName().equals(capability.getName()))
```

Comment:
Updated the variable name to expCapability.
dd6a3eae5a4d5994bc1f5c78582bd06c45fb4382